File: microservers/immobilierservice/src/main/java/io/multi/immobilierservice/domain/Photo.java
package io.multi.immobilierservice.domain;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.OffsetDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Photo {

    private Long photoId;
    private String photoUuid;
    private Long proprieteId;
    private String url;

    private String urlThumbnail;
    private String objectKey;
    private String objectKeyThumbnail;
    private Integer ordreAffichage;
    private Boolean estCouverture;
    private Long tailleOctets;
    private String typeMime;
    private Integer largeur;
    private Integer hauteur;
    private OffsetDateTime createdAt;

    /**
     * URL renvoyée aux clients : route RELATIVE vers le reverse-proxy gateway
     * (cf. PhotoResource#servePhoto, Phase 13b). Permet aux clients web/mobile
     * de préfixer avec leur propre apiBaseUrl (gateway public en prod, host
     * loopback en dev) au lieu d'hériter de l'URL MinIO {@code localhost:9100}
     * stockée historiquement en BD — qui n'est joignable que dans le réseau
     * Docker du backend.
     *
     * <p>Override le getter Lombok @Data. Le champ {@link #url} reste lu/écrit
     * par les row-mappers et les uploads (référence historique), mais n'est
     * plus exposé tel quel via JSON. Fallback sur le champ si {@code photoUuid}
     * est absent (ne devrait jamais arriver pour une photo persistée).
     */
    public String getUrl() {
        if (photoUuid == null || photoUuid.isBlank()) return url;
        return "/immo/photos/" + photoUuid;
    }

    /** Idem {@link #getUrl()}, variante thumbnail via query param. */
    public String getUrlThumbnail() {
        if (photoUuid == null || photoUuid.isBlank()) return urlThumbnail;
        return "/immo/photos/" + photoUuid + "?thumb=true";
    }
}

File: mobile_app/lib/features/immo/models/photo.dart
import '../../../config/app_config.dart';

/// Photo d'une propriété immobilière. Modélise le payload retourné par
/// le backend dans `Propriete.photos[]` (et `photoCouverture`) sur la
/// fiche détail (`GET /immo/proprietes/{uuid}`). Vide en search.
///
/// Champs backend ignorés en 15.2b (à ajouter si un usage UI les requiert) :
/// photoId, proprieteId, objectKey, objectKeyThumbnail, tailleOctets,
/// typeMime, largeur, hauteur, createdAt.
class Photo {
  final String photoUuid;
  final String url;
  final String urlThumbnail;
  final int ordreAffichage;
  final bool estCouverture;

  const Photo({
    required this.photoUuid,
    required this.url,
    required this.urlThumbnail,
    required this.ordreAffichage,
    required this.estCouverture,
  });

  factory Photo.fromJson(Map<String, dynamic> json) => Photo(
        photoUuid: json['photoUuid'] as String,
        url: _absolutize(json['url'] as String),
        urlThumbnail: _absolutize(json['urlThumbnail'] as String),
        ordreAffichage: json['ordreAffichage'] as int? ?? 0,
        estCouverture: json['estCouverture'] as bool? ?? false,
      );

  /// Si le backend renvoie une URL relative (`/immo/photos/{uuid}`), la
  /// préfixe avec l'apiBaseUrl du client (gateway public en prod, host
  /// loopback en dev). Si déjà absolue (legacy MinIO localhost:9100), la
  /// laisse telle quelle — CachedNetworkImage échouera mais affichera son
  /// errorWidget proprement.
  static String _absolutize(String url) {
    if (url.startsWith('http://') || url.startsWith('https://')) {
      return url;
    }
    final base = AppConfig.apiBaseUrl;
    final trimmedBase = base.endsWith('/') ? base.substring(0, base.length - 1) : base;
    return trimmedBase + url;
  }
}
